q2_metadata/normalization/_norm_errors.py

```diff
@@ -0,0 +1,204 @@
+
+    def __str__(self):
+        message = self.generic_message(self.rule, self.variable, self.value)
+        return '%s -> %s' % (message, self.message)
+
+
+class RemapError(Error):
+    """Exception raised for errors in the
+    formatting of the "remap values" rules.
+    """
+    def __init__(self, variable, value):
+        self.rule = 'remap'
+        self.variable = variable
+        self.value = value
+        self.message = self.get_message()
+
+    def get_message(self):
+        if not isinstance(self.value, dict):
+            return 'is not a dictionary'
+        str_values = [x for x, y in self.value.items() if isinstance(y, str)]
+        if len(str_values) != len(self.value.keys()):
+            return 'not all dictionary values are strings'
+
+    def __str__(self):
+        message = self.generic_message(self.rule, self.variable, self.value)
+        return '%s -> %s' % (message, self.message)
+
+
+class ValidationError(Error):
+    """Exception raised for errors in the
+    formatting of the "validation values" rules.
+    """
+    def __init__(self, variable, value):
+        self.rule = 'validation'
+        self.variable = variable
+        self.value = value
+        self.message = self.get_message()
+
+    def get_message(self):
+        possible_validations = {
+            'force_to_blank_if': {'is_null'}
+        }
+        if not isinstance(self.value, dict):
+            return 'is not a dictionary'
+        rule_value_set = set(self.value)
+        accepted_set = rule_value_set & set(possible_validations)
+        non_accepted_set = sorted(rule_value_set ^ accepted_set)
+        if len(non_accepted_set):
+            return 'inacceptable validations: %s' % ', '.join(non_accepted_set)
+        possible_validations_2 = possible_validations['force_to_blank_if']
+        for rule_2, rule_value_2 in self.value['force_to_blank_if'].items():
+            if not isinstance(rule_value_2, list):
+                return 'variables not in a list for condition: %s' % rule_2
+            if rule_2 not in possible_validations_2:
+                return 'inacceptable condition: %s' % rule_2
+
+    def __str__(self):
+        message = self.generic_message(self.rule, self.variable, self.value)
+        return '%s -> %s' % (message, self.message)
+
+
+class NormalizationError(Error):
+    """Exception raised for errors in the
+    formatting of the "validation values" rules.
+    """
+    def __init__(self, variable, value):
+        self.rule = 'normalization'
+        self.variable = variable
+        self.value = value
+        self.message = self.get_message()
+
+    def get_message(self):
+        if not isinstance(self.value, dict):
+            return 'is not a dictionary'
+        possible_normalizations = {'maximum', 'mininum', 'gated_value'}
+        rule_value_set = set(self.value)
+        impossible = sorted(rule_value_set ^ (rule_value_set & possible_normalizations))
```

Comment:
Why are some of these errors doing validation?

Reply:
I am not sure I understand you question.

Note that I was told to use the rules originally defined by Daniel and Rebecca. This rule was an originally defined rule, called "validation": it already corresponded to a dict with keys "maximum", "mininum", "gated_value".

Reply:
I think I get your question. The error is in this instance re-performing a check that failed (because it fired this exception). Here, I re-do the check this time in order to collect the actual terms that are wrong, and thus to chow them to the user as a help.

Again, we may decide to collect everything on the way and through all rules and then return a full summary instead of raising / crashing at the first problem: **this is something I would really appreciate strategic guidance about** - what do you prefer?

Reply:
errors now do not try to re-evaluate/catch the issue - it is done and collected as part of the check_rule() method of class RulesCollection()